dali/pipeline/data/tensor_vector.h

```diff
@@ -0,0 +1,182 @@
+    return tensors_.end();
+  }
+
+  auto end() const noexcept {
+    return tensors_.end();
+  }
+
+  auto cend() const noexcept {
+    return tensors_.cend();
+  }
+
+  auto size() const noexcept {
+    return tensors_.size();
+  }
+
+  kernels::TensorListShape<> shape() const {
+    if (tensors_.empty()) {
+      return {};
+    }
+    kernels::TensorListShape<> result(tensors_.size(), tensors_[0]->ndim());
+    for (size_t i = 0; i < tensors_.size(); i++) {
+      result.set_tensor_shape(i, tensors_[i]->shape());
+    }
+    return result;
+  }
+
+  DLL_PUBLIC inline void Resize(const kernels::TensorListShape<> &new_shape) {
+    // N.B. There probably is nothing wrong with adjusting batchsize for give TensorVector
+    // (sparse tensor list), but the semantics of what type and pinned status should
+    // the new elements have or having some of them allocated and the new ones not
+    // complicates the logic even further so we disallow it
+    DALI_ENFORCE(tensors_.empty() || static_cast<int>(tensors_.size()) == new_shape.size(),
+                 "Changing the batch size is prohibited. It should be set once.");
+    if (tensors_.empty()) {
+      allocate_tensors(new_shape.size());
+    }
+    for (int i = 0; i < new_shape.size(); i++) {
+      tensors_[i]->Resize(new_shape[i]);
+    }
+  }
+
+  inline void set_type(const TypeInfo &new_type) {
+    type_ = new_type;
+    for (auto t : tensors_) {
+      t->set_type(new_type);
+    }
+  }
+
+  inline TypeInfo type() const {
+    // TODO(klecki): do we enforce the same type between elements, or prohibit mixing APIs?
+    return type_;
+  }
+
+  inline void set_pinned(const bool pinned) {
+    // Store the value, in case we pin empty vector and later call Resize
+    pinned_ = pinned;
+    for (auto &t : tensors_) {
+      t->set_pinned(pinned);
+    }
+  }
+
+  inline bool is_pinned() const {
+    // TODO(klecki): do we enforce the same pin status between elements, or prohibit mixing APIs?
+    bool pinned = true;
+    for (auto &t : tensors_) {
+      pinned = pinned && t->is_pinned();
+    }
+    return pinned_ && pinned;
+  }
+
+  /// @brief Reserved memory size is divided between all elements
+  inline void reserve(size_t new_num_bytes) {
+    reserved_batch_memory = new_num_bytes;
+    // Won't do the division by 0 if there are no elements
+    for (auto t : tensors_) {
+      t->reserve(new_num_bytes / tensors_.size());
+    }
+  }
+
+  inline void reserve(size_t new_num_bytes, int batch_size) {
+    DALI_ENFORCE(tensors_.empty() || static_cast<int>(tensors_.size()) == batch_size,
+                 "Changing the batch size is prohibited. It should be set once.");
+    if (tensors_.empty()) {
+      allocate_tensors(batch_size);
+    }
+    for (auto t : tensors_) {
+      t->reserve(new_num_bytes);
+    }
+  }
+
+ private:
+  void allocate_tensors(int batch_size) {
+    DALI_ENFORCE(tensors_.empty(), "Changing the batch size is prohibited. It should be set once.");
```

Comment:
Why? It's relatively easy to extend or truncate the vector...

Reply:
I think that it will make this even more complex in regards to order of pinning/reserving/resizing etc. If we move to something simpler (non-contigues TL with pointers instead of offsets?) we can allow for this.